### PR TITLE
refactor(ephemeral_file): bring back pre-warming of PS page cache on write

### PR DESCRIPTION
**Pull request by problame**
Parent PR https://github.com/neondatabase/neon/pull/7484 is showing a performance regression in `test_bulk_ingest.py`.

I tracked it down to the removal of PS page cache pre-warming: the InMemoryLayer=>L0 writing path seems to benefit from PS page cache for its reads.

So, this PR adds back the pre-warming.
We don't funnel through RequestContext down to the VirtualFile level yet, so, I use a `static` request context.

See https://neondb.slack.com/archives/C033RQ5SPDH/p1713890189520339?thread_ts=1713433920.576379&cid=C033RQ5SPDH for more details on the perf investigation.

### CODE REVIEW

File: pageserver/src/tenant/ephemeral_file/page_caching.rs [hunk 1]
```diff
@@ -96,10 +103,116 @@ impl Drop for RW {
...
+impl crate::virtual_file::owned_buffers_io::write::OwnedAsyncWriter for PreWarmingWriter {
+    async fn write_all<
+        B: tokio_epoll_uring::BoundedBuf<Buf = Buf>,
+        Buf: tokio_epoll_uring::IoBuf + Send,
+    >(
+        &mut self,
+        buf: B,
+    ) -> std::io::Result<(usize, B::Buf)> {
+        let buf = buf.slice(..); // the resulting Slice derefs to &[u8] of [0..buf.bytes_init()]
+        let saved_bounds = buf.bounds(); // save for reconstructing buf from iobuf after the IO is done
```
Comment (VladLazar) [inline pageserver/src/tenant/ephemeral_file/page_caching.rs:139]: The resulting Slice does indeed deref to `&[u8] of [0..buf.bytes_init()]`, but the bounds will be `0..buf.bytes_total()`. That's fine for `zero_padded::Buffer` (`PreWarmingWriter` is only used with this buf type), but I'm not sure what happens if the buffer were to change.

Comment (VladLazar) [reply pageserver/src/tenant/ephemeral_file/page_caching.rs:139]: In C++ I'd add a `static_assert` to ensure that the buffer type of `PrewarmingWriter` is `zero_padded::Buffer`, but I'm not sure what the rust equivalent is.

Comment (VladLazar) [reply pageserver/src/tenant/ephemeral_file/page_caching.rs:139]: Actually, I think `0..buf.bytes_total()` is what you want for reconstructing the buffer.

**Review by VladLazar — COMMENTED** [review]
Looks correct. More of a meta point, but I feel like these interfaces have grown too flexible and we're discovering that it's only safe to plug them in certain ways.

File: pageserver/src/tenant/ephemeral_file/page_caching.rs (hunk 1 above)
Comment (problame) [reply pageserver/src/tenant/ephemeral_file/page_caching.rs:139]: Maybe the wording is wrong.

What I want to achieve here is the following: after the IO returns, I need the same &[u8] slice as what was effectively written to the VirtualFile, such that I can pre-warm the page cache with it.

Will try to improve the wording in the comments.

File: pageserver/src/tenant/ephemeral_file/zero_padded_read_write/flush_detector.rs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+use crate::virtual_file::owned_buffers_io::write::Buffer;
```
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/tenant/ephemeral_file/zero_padded_read_write/flush_detector.rs:1]: Nope :D

File: pageserver/src/tenant/ephemeral_file/page_caching.rs [hunk 3]
```diff
@@ -96,10 +103,116 @@ impl Drop for RW {
...
+            Some(buf.to_vec())
+        } else {
+            None
+        };
+        let buflen = buf.len();
+        assert_eq!(
+            buflen % PAGE_SZ,
+            0,
+            "{buflen} ; we know TAIL_SZ is a PAGE_SZ multiple, and write_buffered_borrowed is used"
+        );
```
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/tenant/ephemeral_file/page_caching.rs:150]: Agree, I don't like it either. Perhaps we should duplicate the BufferedWriter , one for write_buffered_borrowed, one for write_buffered. Not in this PR though.

File: pageserver/src/tenant/ephemeral_file/zero_padded_read_write/flush_detector.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (problame) [reply pageserver/src/tenant/ephemeral_file/zero_padded_read_write/flush_detector.rs]: 3b54288e7d015996ac8ca727bf79245a375cf1ab

File: pageserver/src/tenant/ephemeral_file/page_caching.rs (hunk 1 above)
Comment (problame) [reply pageserver/src/tenant/ephemeral_file/page_caching.rs]: WDYT of d928ad67a390351c4d6b0d45ef88b57765813eaa

Comment (VladLazar) [reply pageserver/src/tenant/ephemeral_file/page_caching.rs]: Yeah, makes sense. This thread was more of a train of thought while reviewing as opposed to something actionable.

**Pull request merged**